File: test/21招新h.py
from cmath import sqrt
from math import ceil

p = [i for i in range(1010)]

def find(x):
    if p[x] != x:
        p[x] = find(p[x])
    return p[x]

n = int(input())
point = []
for i in range(n):
    x, y, v = map(int, input().split())
    point.append([x, y, v])
seq = []
for i in range(n - 1):
    for j in range(i+1, n):
        dis = 0
        if (point[i][2] + point[j][2]) == 0:
            dis = 1e9
        else:
            dis = sqrt(pow(point[i][0] - point[j][0], 2) + pow(point[i][1] - point[j][1], 2)).real / (point[i][2] + point[j][2])
        seq.append([i, j, dis])

seq.sort(key=lambda x:x[2])
cnt = 0
ans = [0 for i in range(1010)]
for i in seq:
    if find(i[0]) == find(i[1]):
        continue
    p[find(i[0])] = find(i[1])
    tmp = ceil(i[2])
    for j in range(cnt, min(tmp, 1005)):
        ans[j] = n
    n -= 1
    cnt = tmp
for i in range(cnt, 1005):
    ans[i] = 1
q = int(input())
for i in range(q):
    t = int(input())
    print(ans[t])

File: test/tmp/21招新h.cpp
#include <cmath>
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
const int N = 1e3+10;
typedef pair<pair<int, int>, int> pii1;
typedef pair<pair<int, int>, double> pii2;
int p[N], ans[N];
int find(int x)
{
    if(p[x] != x) p[x] = find(p[x]);
    return p[x];
}
int main()
{
    int n;
    cin >> n;
    vector<pii1> a;
    for(int i=1;i<=N-5;i++) p[i] = i;
    for (int i = 0; i < n; i++)
    {
        int x, y, v;
        scanf("%d %d %d", &x, &y, &v);
        a.push_back({{x, y}, v});
    }
    vector<pii2> b;
    for (int i = 0; i < n - 1; i++)
    {
        for (int j = i + 1; j < n; j++)
        {
            int x1 = a[i].first.first, x2 = a[j].first.first;
            int y1 = a[i].first.second, y2 = a[j].first.second;
            double t = sqrt((x1-x2)*(x1-x2)+(y1-y2)*(y1-y2)) / (a[i].second + a[j].second);
            b.push_back({{i, j}, t});
        }
        
    }
    sort(b.begin(), b.end(), [](pii2 a, pii2 b){
        return a.second < b.second;
    });
    // for(auto i:b) cout << i.first.first << ' ' << i.first.second << ' ' << i.second << endl;
    int num = n;
    int cnt = 0;
    for(auto i:b)
    {
        if(find(i.first.first) == find(i.first.second)) continue;
        else
        {
            p[find(i.first.first)] = find(i.first.second);
            int tmp = ceil(i.second);
            tmp = min(tmp, 1005);
            for(int j=cnt;j<tmp;j++) ans[j] = num;
            num --;
            cnt = tmp;
        }
    }
    
    for(int i=cnt;i<N;i++) ans[i] = num;
    int q;
    cin >> q;
    while (q--)
    {
        int t;
        cin >> t;
        cout << ans[t] << endl;
    }
    
    return 0;
}